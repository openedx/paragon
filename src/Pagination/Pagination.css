/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-typography-pagination-line-height: 1.5rem;
  --pgn-typography-pagination-font-size-sm: .875rem;
  --pgn-spacing-pagination-margin-x: .5rem;
  --pgn-spacing-pagination-padding-x-lg: 1.5rem;
  --pgn-spacing-pagination-padding-x-sm: .6rem;
  --pgn-spacing-pagination-padding-x-base: 1rem;
  --pgn-spacing-pagination-padding-y-lg: .75rem;
  --pgn-spacing-pagination-padding-y-sm: .8rem;
  --pgn-spacing-pagination-padding-y-base: .625rem;
  --pgn-size-pagination-focus-outline: 0;
  --pgn-size-pagination-focus-border-width: .125rem;
  --pgn-size-pagination-toggle-border-sm: .25rem;
  --pgn-size-pagination-toggle-border-base: .3125rem;
  --pgn-size-pagination-reduced-dropdown-min-width: 6rem;
  --pgn-size-pagination-reduced-dropdown-max-height: 60vh;
  --pgn-size-pagination-secondary-height-sm: 2.25rem;
  --pgn-size-pagination-secondary-height-base: 2.75rem;
  --pgn-size-pagination-icon-height: 2.25rem;
  --pgn-size-pagination-icon-width: 2.25rem;
  --pgn-size-pagination-border-radius-lg: var(--pgn-size-border-radius-lg);
  --pgn-size-pagination-border-radius-sm: var(--pgn-size-border-radius-sm);
  --pgn-size-pagination-border-width: var(--pgn-size-border-width);
}
