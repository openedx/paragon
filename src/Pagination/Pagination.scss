@import "variables";
@import "~bootstrap/scss/pagination";

.pagination {
  align-items: center;

  .dropdown {
    z-index: 4;
  }
}

%pagination-icon-button-right {
  border-top-right-radius: 50%;
  border-bottom-right-radius: 50%;
}

%pagination-icon-button-left {
  border-top-left-radius: 50%;
  border-bottom-left-radius: 50%;
}

.pagination-icon-button-right {
  @extend %pagination-icon-button-right;
}

.pagination-icon-button-left {
  @extend %pagination-icon-button-left;
}

.pagination-default {
  .page-link {
    &.previous .pgn__icon {
      margin-inline-start: 0;
      margin-inline-end: $pagination-margin-x;
    }

    &.next .pgn__icon {
      margin-inline-start: $pagination-margin-x;
      margin-inline-end: 0;
    }
  }

  .page-item {
    &:first-child .page-link {
      [dir="rtl"] & {
        border-radius: 0 $pagination-border-radius-lg $pagination-border-radius-lg 0;
      }
    }

    &:last-child .page-link {
      [dir="rtl"] & {
        border-radius: $pagination-border-radius-lg 0 0 $pagination-border-radius-lg;
      }
    }
  }
}

.page-link {
  border: none;

  &.btn-primary:not(:disabled):not(.disabled):focus {
    background-color: $pagination-bg;
    color: $pagination-focus-color-text;
  }

  &:focus {
    box-shadow: none;
  }

  &.btn-primary:focus::before {
    border: $pagination-focus-border-width solid $pagination-focus-color;

    @include button-size($btn-padding-y, $btn-padding-x, $btn-font-size, $btn-line-height, $btn-border-radius);
  }

  div {
    display: flex;
  }

  [dir="rtl"] & {
    svg {
      transform: scale(-1);
    }
  }

  &:focus::before,
  &.focus::before {
    border-radius: 0;
  }
}

.page-item {
  > .btn {
    transition: none;
    line-height: $pagination-line-height;
  }

  &.active .page-link.btn-primary:not(:disabled):not(.disabled):focus {
    background-color: $pagination-focus-color;
    color: $pagination-bg;
  }
}

.pagination-small {
  .page-link {
    font-size: $pagination-font-size-sm;
    line-height: $pagination-line-height;
    padding: .375rem .78rem;

    &.previous,
    &.next {
      padding: 0 $pagination-padding-y;
      line-height: $pagination-secondary-height-sm;

      div {
        display: flex;
        align-items: center;
      }
    }
  }

  &:not(.pagination-default) {
    .page-link {
      &.previous,
      &.next {
        padding: 0;
        width: $pagination-icon-width;
      }
    }
  }
}

.pagination-secondary {
  button.next,
  button.previous {
    height: $pagination-secondary-height;
    padding: 0 $pagination-padding-y;
  }

  &.pagination-small {
    button.next,
    button.previous {
      height: $pagination-secondary-height-sm;
      line-height: $pagination-line-height;
    }
  }

  .page-item:first-child .page-link {
    @extend %pagination-icon-button-left;
  }

  .page-item:last-child .page-link {
    @extend %pagination-icon-button-right;
  }
}

.pagination-inverse {
  %dark-styles {
    background-color: transparent;
    color: $white;
  }

  .pgn__dark-styles {
    @extend %dark-styles;
  }

  .page-item {
    &.disabled .page-link {
      @extend %dark-styles;
    }

    &.active button.page-link {
      background-color: $pagination-bg;
      color: $pagination-color;
    }

    button.page-link {
      @extend %dark-styles;

      &:focus {
        box-shadow: none;
      }
    }

    &:not(.active):focus {
      box-shadow: $level-1-box-shadow;
    }
  }

  .page-link {
    &:focus::before,
    &.focus::before {
      display: none;
    }
  }

  .dropdown {
    .btn-tertiary {
      color: $pagination-color-inverse;

      &::after {
        border-top: $pagination-toggle-border solid $pagination-color-inverse;
      }

      &:active,
      &:hover {
        background-color: transparent;
      }

      &:not(:disabled):not(.disabled):active {
        color: $pagination-color-inverse;
      }
    }
  }

  .show > .dropdown-toggle {
    background-color: transparent;
  }
}

.pgn__reduced-pagination-dropdown {
  overflow-y: auto;
  max-height: $pagination-reduced-dropdown-max-height;
  min-width: $pagination-reduced-dropdown-min-width;

  a {
    text-align: center;
  }
}

.pagination-reduced {
  .dropdown-toggle::after {
    width: 0;
    height: 0;
    border-left: $pagination-toggle-border solid transparent;
    border-right: $pagination-toggle-border solid transparent;
    border-top: $pagination-toggle-border solid $gray-700;
    transform: rotate(0);
    inset-inline-start: .5rem;
    top: 0;
    margin-inline-end: 1rem;
  }

  button.next,
  button.previous {
    height: $pagination-secondary-height;
    padding: 0 $pagination-padding-y;
  }

  &.pagination-small {
    .btn.dropdown-toggle {
      font-size: $pagination-font-size-sm;

      &::after {
        border-left-width: $pagination-toggle-border-sm;
        border-right-width: $pagination-toggle-border-sm;
        border-top-width: $pagination-toggle-border-sm;
      }
    }

    button.previous,
    button.next {
      line-height: $pagination-icon-height;
      height: $pagination-icon-height;
    }
  }

  .page-item:first-child .page-link {
    @extend %pagination-icon-button-left;
  }

  .page-item:last-child .page-link {
    @extend %pagination-icon-button-right;
  }
}

.pagination-minimal {
  .page-item:first-child {
    margin-inline-end: .3rem;
  }

  button.next,
  button.previous {
    padding: $pagination-padding-y;
    height: $pagination-secondary-height;
  }

  &.pagination-small {
    button.next,
    button.previous {
      padding: 0 $pagination-padding-y;
      height: $pagination-secondary-height-sm;
    }
  }

  .page-item:first-child .page-link {
    @extend %pagination-icon-button-left;
  }

  .page-item:last-child .page-link {
    @extend %pagination-icon-button-right;
  }
}
