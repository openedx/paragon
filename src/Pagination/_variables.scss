// Pagination

$pagination-padding-y:              .625rem !default;
$pagination-margin-x:               .5rem !default;
$pagination-line-height:            1.5rem !default;
$pagination-font-size-sm:           .875rem !default;

$pagination-icon-width:             2.25rem !default;
$pagination-icon-height:            2.25rem !default;
$pagination-toggle-border:          .3125rem !default;
$pagination-toggle-border-sm:       .25rem !default;
$pagination-secondary-height:       2.75rem !default;
$pagination-secondary-height-sm:    2.25rem !default;

$pagination-dropdown-color-inverse: $white !default;
$pagination-border-width:           $border-width !default;

$pagination-reduced-dropdown-max-height: 60vh !default;
$pagination-reduced-dropdown-min-width: 6rem !default;
