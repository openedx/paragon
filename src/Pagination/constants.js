/* eslint-disable import/prefer-default-export */
export const ELLIPSIS = '...';
