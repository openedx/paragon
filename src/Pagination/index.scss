@import "variables";

%pagination-icon-button-right {
  border-top-right-radius: 50%;
  border-bottom-right-radius: 50%;
}

%pagination-icon-button-left {
  border-top-left-radius: 50%;
  border-bottom-left-radius: 50%;
}

.pagination {
  display: flex;
  margin: 0;

  .dropdown {
    z-index: 4;
  }

  .page-of-count {
    margin: 0 .5rem;
    border: 0;
  }

  .page-item {
    &:first-child .page-link {
      margin-left: 0;

      @include border-left-radius($border-radius);
    }

    &:last-child .page-link {
      @include border-right-radius($border-radius);
    }

    &:first-child .btn-icon.page-link {
      @extend %pagination-icon-button-left;
    }

    &:last-child .btn-icon.page-link {
      @extend %pagination-icon-button-right;
    }

    &.active .page-link {
      z-index: 3;
    }

    > .btn {
      transition: none;
      line-height: $pagination-line-height;
    }
  }

  @include list-unstyled();
  @include border-radius();

  &-small {
    .page-link {
      font-size: $pagination-font-size-sm;
      line-height: $pagination-line-height;
      padding: .375rem .78rem;

      &.previous,
      &.next {
        padding: 0 $pagination-padding-y;
        line-height: $pagination-secondary-height-sm;

        div {
          display: flex;
          align-items: center;
        }
      }
    }

    &:not(.pagination-default) .page-link {
      &.previous,
      &.next {
        padding: 0;
        width: $pagination-icon-width;
      }
    }
  }

  &-secondary {
    button.next,
    button.previous {
      height: $pagination-secondary-height;
      padding: 0 $pagination-padding-y;
    }

    &.pagination-small {
      button.next,
      button.previous {
        height: $pagination-secondary-height-sm;
        line-height: $pagination-line-height;
      }
    }
  }

  .ellipsis {
    border: 0;
    margin-left: 0;
  }

  &-inverse {
    .ellipsis {
      color: $white;
    }

    .dropdown .dropdown-toggle::after {
      border-top: $pagination-toggle-border solid $pagination-dropdown-color-inverse;
    }
  }

  &-reduced {
    &-dropdown-menu {
      overflow-y: auto;
      max-height: $pagination-reduced-dropdown-max-height;
      min-width: $pagination-reduced-dropdown-min-width;

      a {
        text-align: center;
      }
    }

    .dropdown-toggle::after {
      width: 0;
      height: 0;
      border-left: $pagination-toggle-border solid transparent;
      border-right: $pagination-toggle-border solid transparent;
      border-top: $pagination-toggle-border solid $gray-700;
      transform: rotate(0);
      inset-inline-start: .5rem;
      top: 0;
      margin-inline-end: 1rem;
    }

    button.next,
    button.previous {
      height: $pagination-secondary-height;
      padding: 0 $pagination-padding-y;
    }

    &.pagination-small {
      .btn.dropdown-toggle {
        font-size: $pagination-font-size-sm;

        &::after {
          border-left-width: $pagination-toggle-border-sm;
          border-right-width: $pagination-toggle-border-sm;
          border-top-width: $pagination-toggle-border-sm;
        }
      }

      button.previous,
      button.next {
        line-height: $pagination-icon-height;
        height: $pagination-icon-height;
      }
    }
  }

  &-minimal {
    .page-item:first-child {
      margin-inline-end: .3rem;
    }

    button.next,
    button.previous {
      padding: $pagination-padding-y;
      height: $pagination-secondary-height;
    }

    &.pagination-small {
      button.next,
      button.previous {
        padding: 0 $pagination-padding-y;
        height: $pagination-secondary-height-sm;
      }
    }
  }
}

.page-link {
  border: none;
  margin-left: -$pagination-border-width;

  &:focus {
    z-index: 3;
  }

  div {
    display: flex;
  }

  [dir="rtl"] & {
    svg {
      transform: scale(-1);
    }
  }
}
