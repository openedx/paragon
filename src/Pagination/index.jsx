import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import ReducedPagination from './ReducedPagination';
import MinimalPagination from './MinimalPagination';
import DefaultPagination from './DefaultPagination';
import { PaginationContextProvider } from './PaginationContext';
import { PAGINATION_VARIANTS } from './constants';
import { ScreenReaderText } from './subcomponents';

import { greaterThan } from '../utils/propTypes';
import { ChevronLeft, ChevronRight } from '../../icons';

function Pagination(props) {
  const {
    invertColors,
    variant,
    size,
    paginationLabel,
    className,
  } = props;

  const renderPaginationComponent = () => {
    if (variant === PAGINATION_VARIANTS.reduced) {
      return <ReducedPagination />;
    }

    if (variant === PAGINATION_VARIANTS.minimal) {
      return <MinimalPagination />;
    }

    return <DefaultPagination />;
  };

  return (
    <PaginationContextProvider {...props}>
      <nav
        aria-label={paginationLabel}
        className={classNames(className, {
          [`pagination-${variant}`]: variant,
          'pagination-inverse': invertColors,
          'pagination-small': size !== PAGINATION_VARIANTS.default,
        })}
      >
        <ScreenReaderText />
        {renderPaginationComponent()}
      </nav>
    </PaginationContextProvider>
  );
}

Pagination.propTypes = {
  /**
   * Specifies a callback function that is executed when the
   * user selects a page button or the `Previous`/`Next` buttons. For example:
   *
   * ```jsx
   *  <Pagination onPageSelect={(pageNumber) => { console.log(pageNumber); } />
   * ```
   */
  onPageSelect: PropTypes.func.isRequired,
  /** Specifies the total number of pages in the `Pagination` component. */
  pageCount: PropTypes.number.isRequired,
  /** Specifies the `aria-label` for the `<nav>` element that wraps the pagination button list. */
  paginationLabel: PropTypes.string.isRequired,
  /**
   * Specifies the labels to use for the `Previous`/`Next`
   * buttons as well as the various parts of `aria-label`
   * on the page buttons for accessibility. All button labels
   * accept both string or elements. The button label options are as follows:
   *
   * `previous`: Text for the `Previous` button;
   *
   * `next`: Text for the `Next` button;
   *
   * `page`: Text in the `aria-label` on page buttons to describe the button (e.g., "**Page** 1");
   *
   * `currentPage`: Text to depict the selected page in the `aria-label`
   * on page buttons (e.g., "Page 1, **Current Page**");
   *
   * `pageOfCount`: Text that separates the current page and total page count
   * for the mobile UI (e.g., "Page 1 **of** 20").
   *
   * The default is:
   * ```javascript
   * {
   *   previous: 'Previous',
   *   next: 'Next',
   *   page: 'Page',
   *   currentPage: 'Current Page',
   *   pageOfCount: 'of',
   * }
   * ```
   */
  buttonLabels: PropTypes.shape({
    previous: PropTypes.string,
    next: PropTypes.string,
    page: PropTypes.string,
    currentPage: PropTypes.string,
    pageOfCount: PropTypes.string,
  }),
  /** Specifies any class name(s) for the `Pagination` component. The default is an empty string. */
  className: PropTypes.string,
  /** specifies the page that the `Pagination` component will automatically select. The default is `1`. */
  currentPage: PropTypes.number,
  /**
   * Specifies the number of page buttons to display in between the `Previous`
   * and `Next` buttons. This number also includes any ellipses in the total count.
   * Also, to ensure that at least one clickable page button is shown when both ellipses
   * are displayed, this value must be greater than `4`. The default is `7`.
   */
  maxPagesDisplayed: greaterThan(4),
  /**
   * Specifies icons used to indicate previous and next page. Can be an element,
   * string, symbol, etc. Default is chevrons rendered using fa-css.
   */
  icons: PropTypes.shape({
    leftIcon: PropTypes.elementType,
    rightIcon: PropTypes.elementType,
  }),
  variant: PropTypes.oneOf(['default', 'secondary', 'reduced', 'minimal']),
  invertColors: PropTypes.bool,
  size: PropTypes.oneOf(['default', 'small']),
  initialPage: PropTypes.number,
};

Pagination.defaultProps = {
  icons: {
    leftIcon: ChevronLeft,
    rightIcon: ChevronRight,
  },
  buttonLabels: {
    previous: 'Previous',
    next: 'Next',
    page: 'Page',
    currentPage: 'Current Page',
    pageOfCount: 'of',
  },
  className: undefined,
  initialPage: 1,
  currentPage: undefined,
  maxPagesDisplayed: 7,
  variant: 'default',
  invertColors: false,
  size: 'default',
};

export default Pagination;
export * from './constants';
