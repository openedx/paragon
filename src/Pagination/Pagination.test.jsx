import React from 'react';
import { mount } from 'enzyme';
import { Context as ResponsiveContext } from 'react-responsive';
import { act } from 'react-dom/test-utils';
import breakpoints from '../utils/breakpoints';
import Pagination from './index';
import Dropdown from '../Dropdown';

const baseProps = {
  state: { pageIndex: 1 },
  paginationLabel: 'pagination navigation',
  pageCount: 5,
  onPageSelect: () => {},
};

describe('<Pagination />', () => {
  it('renders', () => {
    const props = {
      ...baseProps,
    };
    const wrapper = mount(<Pagination {...props} />);
    expect(wrapper.exists()).toEqual(true);
  });

  it('renders screen reader section', () => {
    const buttonLabels = {
      previous: 'Anterior',
      next: 'Siguiente',
      page: 'P치gina',
      currentPage: 'P치gina actual',
      pageOfCount: 'de',
    };
    const props = {
      ...baseProps,
      buttonLabels,
    };
    const wrapper = mount(<Pagination {...props} />);
    expect(
      wrapper.findWhere(node => node.hasClass('sr-only')).text(),
    ).toEqual(`${buttonLabels.page} 1, ${buttonLabels.currentPage}, ${buttonLabels.pageOfCount} ${baseProps.pageCount}`);
  });

  describe('handles currentPage props properly', () => {
    it('overrides state currentPage when props currentPage changes', () => {
      const initialPage = 1;
      const newPage = 2;
      const props = {
        ...baseProps,
        currentPage: initialPage,
      };
      const wrapper = mount(<Pagination {...props} />);
      expect(wrapper.state('currentPage')).toEqual(initialPage);
      wrapper.setProps({
        currentPage: newPage,
      });
      wrapper.update();
      expect(wrapper.state('currentPage')).toEqual(newPage);
    });

    it('does not override state currentPage when props currentPage changes with existing value', () => {
      const currentPage = 2;
      const props = {
        ...baseProps,
        currentPage,
      };
      const wrapper = mount(<Pagination {...props} />);
      expect(wrapper.state('currentPage')).toEqual(currentPage);
      wrapper.setProps({
        currentPage,
      });
      wrapper.update();
      expect(wrapper.state('currentPage')).toEqual(currentPage);
    });
  });

  describe('handles focus properly', () => {
    it('should change focus to next button if previous page is first page', () => {
      const props = {
        ...baseProps,
        currentPage: 2,
      };
      const app = document.createElement('div');
      document.body.appendChild(app);
      const wrapper = mount(<Pagination {...props} />, { attachTo: app });
      wrapper.find('button.previous').simulate('click');
      expect(wrapper.find('button.next').instance()).toEqual(document.activeElement);
    });

    it('should change focus to previous button if next page is last page', () => {
      const props = {
        ...baseProps,
        currentPage: baseProps.pageCount - 1,
      };
      const app = document.createElement('div');
      document.body.appendChild(app);
      const wrapper = mount(<Pagination {...props} />, { attachTo: app });
      wrapper.find('button.next').simulate('click');
      expect(wrapper.find('button.previous').instance()).toEqual(document.activeElement);
    });
  });

  describe('renders required props', () => {
    it('should use correct paginationLabel', () => {
      const paginationLabel = 'pagination navigation';
      const props = {
        ...baseProps,
        paginationLabel,
      };
      const wrapper = mount(<Pagination {...props} />);
      expect(wrapper.find('nav').prop('aria-label')).toEqual(paginationLabel);
    });

    describe('should use correct number of pages', () => {
      it('should show 5 buttons on desktop', () => {
        const wrapper = mount((
          <ResponsiveContext.Provider value={{ width: breakpoints.large.maxWidth }}>
            <Pagination {...baseProps} />
          </ResponsiveContext.Provider>
        ));
        expect(wrapper.findWhere((node) => {
          const isPrevOrNext = node.hasClass('previous') || node.hasClass('next');
          return node.name() === 'button' && !isPrevOrNext;
        })).toHaveLength(5);
      });

      it('should show 1 button on mobile', () => {
        // Use extra small window size to display the mobile version of `Pagination`.
        const wrapper = mount((
          <ResponsiveContext.Provider value={{ width: breakpoints.extraSmall.maxWidth }}>
            <Pagination {...baseProps} />
          </ResponsiveContext.Provider>
        ));
        expect(wrapper.findWhere((node) => {
          const name = node.name();
          const isPrevOrNext = node.hasClass('previous') || node.hasClass('next');
          return name === 'span' && node.hasClass('btn') && !isPrevOrNext;
        })).toHaveLength(1);
      });
    });

    describe('should fire callbacks properly', () => {
      it('should not fire onPageSelect when selecting current page', () => {
        const spy = jest.fn();
        const props = {
          ...baseProps,
          onPageSelect: spy,
        };
        const wrapper = mount((
          <ResponsiveContext.Provider value={{ width: breakpoints.large.maxWidth }}>
            <Pagination {...props} />
          </ResponsiveContext.Provider>
        ));

        wrapper.find('.btn').at(1).simulate('click');
        expect(spy).toHaveBeenCalledTimes(0);
      });

      it('should fire onPageSelect callback when selecting new page', () => {
        const spy = jest.fn();
        const props = {
          ...baseProps,
          onPageSelect: spy,
        };
        const wrapper = mount((
          <ResponsiveContext.Provider value={{ width: breakpoints.large.maxWidth }}>
            <Pagination {...props} />
          </ResponsiveContext.Provider>
        ));

        wrapper.find('.btn').at(2).simulate('click');
        expect(wrapper.state('currentPage')).toEqual(2);
        expect(spy).toHaveBeenCalledTimes(1);

        wrapper.find('.btn').at(3).simulate('click');
        expect(wrapper.state('currentPage')).toEqual(3);
        expect(spy).toHaveBeenCalledTimes(2);
      });
    });
  });

  describe('fires previous and next button click handlers', () => {
    it('previous button onClick', () => {
      const spy = jest.fn();
      const props = {
        ...baseProps,
        onPageSelect: spy,
      };
      const wrapper = mount(<Pagination {...props} />);
      wrapper.setProps({ currentPage: 2 });
      wrapper.update();
      wrapper.find('button.previous').simulate('click');
      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('next button onClick', () => {
      const spy = jest.fn();
      const props = {
        ...baseProps,
        onPageSelect: spy,
      };
      const wrapper = mount(<Pagination {...props} />);
      wrapper.find('button.next').simulate('click');
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });

  describe('renders button label props', () => {
    const buttonLabels = {
      previous: 'Anterior',
      next: 'Siguiente',
      page: 'P치gina',
      currentPage: 'P치gina actual',
      pageOfCount: 'de',
    };

    let wrapper;
    let props;
    /**
     * made a proxy component because setProps can only be used with root component and
     * Responsive Context Provider is needed to mock screen
     * */
    // eslint-disable-next-line react/prop-types
    function Proxy({ currentPage, width }) {
      return (
        <ResponsiveContext.Provider value={{ width }}>
          <Pagination {...props} currentPage={currentPage} />
        </ResponsiveContext.Provider>
      );
    }

    beforeEach(() => {
      props = {
        ...baseProps,
        buttonLabels,
      };
      wrapper = mount(
        <Proxy currentPage={1} width={breakpoints.large.minWidth} />,
      );
    });

    it('uses passed in previous button label', () => {
      expect(wrapper.findWhere(node => (
        node.name() === 'button' && node.hasClass('previous')
      )).prop('aria-label')).toEqual(buttonLabels.previous);

      wrapper.setProps({ currentPage: baseProps.pageCount });
      wrapper.update();

      expect(wrapper.findWhere(node => (
        node.name() === 'button' && node.hasClass('previous')
      )).prop('aria-label')).toEqual(`${buttonLabels.previous}, ${buttonLabels.page} 4`);
    });

    it('uses passed in next button label', () => {
      expect(wrapper.findWhere(node => (
        node.name() === 'button' && node.hasClass('next')
      )).prop('aria-label')).toEqual(`${buttonLabels.next}, ${buttonLabels.page} 2`);

      wrapper.setProps({ currentPage: baseProps.pageCount });
      wrapper.update();

      expect(wrapper.findWhere(node => (
        node.name() === 'button' && node.hasClass('next')
      )).prop('aria-label')).toEqual(buttonLabels.next);
    });

    it('uses passed in page button label', () => {
      wrapper = mount((
        <ResponsiveContext.Provider value={{ width: breakpoints.large.minWidth }}>
          <Pagination {...props} />
        </ResponsiveContext.Provider>
      ));
      expect(wrapper.state('currentPage')).toEqual(1);
      expect(wrapper.find('.btn').at(1).prop('aria-label'))
        .toEqual(`${buttonLabels.page} 1, ${buttonLabels.currentPage}`);
      wrapper = mount((
        <ResponsiveContext.Provider value={{ width: breakpoints.large.minWidth }}>
          <Pagination {...props} currentPage={2} />
        </ResponsiveContext.Provider>
      ));

      expect(wrapper.state('currentPage')).toEqual(2);
      expect(wrapper.find('.btn').at(1).prop('aria-label'))
        .toEqual(`${buttonLabels.page} 1`);

      wrapper = mount(
        <Proxy currentPage={1} width={breakpoints.extraSmall.maxWidth} />,
      );

      expect(wrapper.find('.btn').at(1).prop('aria-label'))
        .toEqual(`${buttonLabels.page} 1, ${buttonLabels.currentPage}, ${buttonLabels.pageOfCount} 5`);
    });

    it('for the reduced variant shows dropdown button with the page count as label', async () => {
      wrapper = mount(<Pagination variant="reduced" {...props} />);

      const dropdown = wrapper.find(Dropdown);
      expect(dropdown.text()).toContain(`${baseProps.state.pageIndex} of ${baseProps.pageCount}`);

      const dropdownButton = wrapper.find('button.dropdown-toggle');
      dropdownButton.simulate('click');
      await act(async () => {
        const dropdownChoices = wrapper.find(Dropdown.Item);
        expect(dropdownChoices.length).toEqual(baseProps.pageCount);
      });
    });

    it('renders only previous and next buttons in minimal variant', () => {
      wrapper = mount(
        <Pagination
          variant="minimal"
          onPageSelect={(pageNumber) => pageNumber}
          pageCount={12}
          paginationLabel="Label"
        />,
      );
      const items = wrapper.find('li.page-item');

      expect(items.length).toEqual(2);
    });

    it('renders chevrons and buttons disabled when pageCount is 1 || 0 for all variants', () => {
      const variantTypes = ['default', 'secondary', 'reduced', 'minimal'];
      // default
      variantTypes.forEach((variantType) => {
        for (let i = 0; i < 3; i++) {
          props = {
            ...baseProps,
            variant: variantType,
            pageCount: i,
          };
          wrapper = mount(<Pagination {...props} />);
          const disabled = wrapper.find('button[disabled=true]');
          expect(props.pageCount).toEqual(i);
          expect(disabled.length).toEqual(i === 2 ? 1 : 2);
        }
      });
    });
  });
});
