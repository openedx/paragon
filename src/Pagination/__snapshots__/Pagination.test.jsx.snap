// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders default variant 1`] = `
<nav
  aria-label="pagination navigation"
  className="pagination-default"
>
  <div
    aria-atomic={true}
    aria-live="polite"
    aria-relevant="text"
    className="sr-only"
  >
    Page 1, Current Page, of 5
  </div>
  <ul
    className="pagination"
  >
    <li
      className="page-item disabled"
    >
      <button
        aria-label="Previous"
        className="previous page-link btn btn-tertiary"
        disabled={true}
        onClick={[Function]}
        tabIndex="-1"
        type="button"
      >
        <span
          className="pgn__icon btn-icon-before"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59Z"
              fill="currentColor"
            />
          </svg>
        </span>
        Previous
      </button>
    </li>
    <li
      className="page-item active"
    >
      <button
        aria-label="Page 1, Current Page"
        className="page-link btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 2"
        className="page-link btn btn-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 3"
        className="page-link btn btn-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 4"
        className="page-link btn btn-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 5"
        className="page-link btn btn-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Next, Page 2"
        className="previous page-link btn btn-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Next
        <span
          className="pgn__icon btn-icon-after"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`<Pagination /> renders with inverse colors 1`] = `
<nav
  aria-label="pagination navigation"
  className="pagination-default pagination-inverse"
>
  <div
    aria-atomic={true}
    aria-live="polite"
    aria-relevant="text"
    className="sr-only"
  >
    Page 1, Current Page, of 5
  </div>
  <ul
    className="pagination"
  >
    <li
      className="page-item disabled"
    >
      <button
        aria-label="Previous"
        className="previous page-link btn btn-inverse-tertiary"
        disabled={true}
        onClick={[Function]}
        tabIndex="-1"
        type="button"
      >
        <span
          className="pgn__icon btn-icon-before"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.41 7.41 14 6l-6 6 6 6 1.41-1.41L10.83 12l4.58-4.59Z"
              fill="currentColor"
            />
          </svg>
        </span>
        Previous
      </button>
    </li>
    <li
      className="page-item active"
    >
      <button
        aria-label="Page 1, Current Page"
        className="page-link btn btn-inverse-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 2"
        className="page-link btn btn-inverse-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 3"
        className="page-link btn btn-inverse-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 4"
        className="page-link btn btn-inverse-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Page 5"
        className="page-link btn btn-inverse-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        aria-label="Next, Page 2"
        className="previous page-link btn btn-inverse-tertiary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Next
        <span
          className="pgn__icon btn-icon-after"
        >
          <svg
            aria-hidden={true}
            fill="none"
            focusable={false}
            height={24}
            role="img"
            viewBox="0 0 24 24"
            width={24}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6-6-6Z"
              fill="currentColor"
            />
          </svg>
        </span>
      </button>
    </li>
  </ul>
</nav>
`;
